src/Compiler/CompileExpr.idr

```diff
@@ -556,43 +562,43 @@ mkArgList i (S k)
           (_ ** ConsArg (MN "arg" i) rec)
 
 data NArgs : Type where
-     User : Name -> List (Closure []) -> NArgs
-     Struct : String -> List (String, Closure []) -> NArgs
+     User : Name -> List (Closure [<]) -> NArgs
+     Struct : String -> List (String, Closure [<]) -> NArgs
      NUnit : NArgs
      NPtr : NArgs
      NGCPtr : NArgs
      NBuffer : NArgs
      NForeignObj : NArgs
-     NIORes : Closure [] -> NArgs
+     NIORes : Closure [<] -> NArgs
 
 getPArgs : {auto c : Ref Ctxt Defs} ->
-           Defs -> Closure [] -> Core (String, Closure [])
+           Defs -> Closure [<] -> Core (String, Closure [<])
 getPArgs defs cl
     = do NDCon fc _ _ _ args <- evalClosure defs cl
              | nf => throw (GenericMsg (getLoc nf) "Badly formed struct type")
-         case reverse (map snd args) of
-              (tydesc :: n :: _) =>
+         case map snd args of
+              (_ :< n :< tydesc) =>
```

Comment:
This looks suspicious. All over the code above the order is changed, I mean, list `a :: b :: cs` becomes `cs :< b :< a`. Here arguments are intensionally reversed, i.e. the opposite (outmost) part of the args was matches. This reversing was removed which looks strange. Looks as if some reversing conversion to list should be used, and the pattern left as is.